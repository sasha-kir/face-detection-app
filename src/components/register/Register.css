.register-form {
  padding: 7px 20px 30px;
}