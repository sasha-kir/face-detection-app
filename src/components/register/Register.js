import React, { Component } from "react";
import "./Register.css";
import { withAlert } from "react-alert";

class Register extends Component {
	constructor(props) {
		super(props);
		this.state = {
			name: "",
			email: "",
			password: ""
		}
	}

	onNameChange = (event) => {
		this.setState({ name: event.target.value });
	}

	onEmailChange = (event) => {
		this.setState({ email: event.target.value });
	}

	onPasswordChange = (event) => {
		this.setState({ password: event.target.value });
	}

	handleEnter = (event) => {
   		if (event.key === "Enter") {
   			this.handleSubmit();
   		}
   	}

	handleSubmit = () => {
		const { name, email, password } = this.state;
		const emailInput = document.getElementById("email-input");

		if (!name || !email || !password) {
			this.props.alert.show("Please fill in all appropriate fields");
		} else if (!emailInput.checkValidity()) {
			this.props.alert.show("Please provide a valid email");
		} else {
			fetch(process.env.REACT_APP_SERVER_URL + "/register", {
				method: "post",
				headers: {'Content-Type': 'application/json'},
				body: JSON.stringify({
					name: name,
					email: email,
					password: password
				})
			})
			  .then(response => {
			  	if (response.status === 200) {
			  		return Promise.resolve(response.json())
			  	} else {
			  		var error = new Error(response.statusText || response.status);
      				error.response = response;
      				return Promise.reject(error)
			  	}
			  })
			  .then(user => {
			      	this.props.loadUser(user);
			        this.props.onRouteChange("home");
			  })
			  .catch(err => {
			  	this.props.alert.error("User with this email already exists");
			  })
		}
	}

	render() {
		const emailPattern = "^[a-zA-Z0-9\\.\\-_]+@[a-z]+\\.[a-z]{2,3}$";
		const { onRouteChange } = this.props;
		return (
			<div className="sign-in-main-div center-content">
				<div className="sign-in-wrapper">
					<form className="register-form sign-in-form">
					  <div className="form-element f4 white">
					    <h1>Registration</h1>
					  </div>

					  <div className="form-element block-cube block-input">
					    <input placeholder="your name" type="text" 
					    	   onChange={this.onNameChange} required />
					    <div className="cube-top"><div className="cube-top-inner"></div></div>
					    <div className="cube-right"><div className="cube-right-inner"></div></div>
					    <div className="cube"><div className="cube-inner"></div></div>
					  </div>

					  <div className="form-element block-cube block-input">
					    <input id="email-input" placeholder="email" type="text"
					           pattern={emailPattern} onChange={this.onEmailChange} required />
					    <div className="cube-top"><div className="cube-top-inner"></div></div>
					    <div className="cube-right"><div className="cube-right-inner"></div></div>
					    <div className="cube"><div className="cube-inner"></div></div>
					  </div>

					  <div className="form-element block-cube block-input">
					    <input placeholder="password" type="password" 
					           onChange={this.onPasswordChange} 
					           onKeyDown={this.handleEnter} required />
					    <div className="cube-top"><div className="cube-top-inner"></div></div>
					    <div className="cube-right"><div className="cube-right-inner"></div></div>
					    <div className="cube"><div className="cube-inner"></div></div>
					  </div>

					  <button className="block-button block-cube grow" 
					  	      type="button" onClick={this.handleSubmit} >
					    <div className="cube-top"><div className="cube-top-inner"></div></div>
					    <div className="cube-right"><div className="cube-right-inner"></div></div>
					    <div className="cube"><div className="cube-inner"></div></div>
					    <div className="login-text button-text">Sign Up</div>
					  </button>

					  <div className="link dim pointer mt4 mb1 f5 white center-content" 
					       onClick={() => onRouteChange("sign-in")}>
					  	Go back
					  </div>
					</form>
				</div>
			</div>
		);
	}
}

export default withAlert()(Register);
