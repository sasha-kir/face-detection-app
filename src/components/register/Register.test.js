import React from 'react';
import { act } from 'react-dom/test-utils';  
import { mount } from 'enzyme';
import nock from 'nock';
import waitUntil from 'async-wait-until';
import { Provider as AlertProvider } from "react-alert";
import AlertTemplate from '../alert-template/react-alert-custom-template';

import Register from './Register';

const loadUser = jest.fn();
const onRouteChange = jest.fn();

describe('Register', () => {
    let component;

    beforeEach(() => {
        component = mount(
            <AlertProvider template={AlertTemplate}>
                <Register onRouteChange={onRouteChange} 
                          loadUser={loadUser} />
            </AlertProvider>);
    });

    afterEach(() => {
        component.unmount();
        nock.cleanAll();
    });

    it('renders correctly', () => {
        const wrapper= mount(<Register />);
        expect(wrapper).toMatchSnapshot();
        wrapper.unmount();
    });
    it('routes to SignIn', () => {
        component.find('#signin-link').simulate('click');
        expect(onRouteChange).toHaveBeenCalledWith('sign-in');
    });
    it('saves name, email and password to state', () => {
        const form = component.find('Register');
        const nameInput = component.find('input#name').at(0);
        const emailInput = component.find('input#email').at(0);
        const passInput = component.find('input#password').at(0);

        nameInput.instance().value = "Amy";
        nameInput.simulate('change');
        expect(form.state('name')).toBe("Amy");

        emailInput.instance().value = "amy@gmail.com";
        emailInput.simulate('change');
        expect(form.state('email')).toBe("amy@gmail.com");

        passInput.instance().value = "secret";
        passInput.simulate('change');
        expect(form.state('password')).toBe("secret");
    });
    it('handles empty fields on submit', () => {
        const form = component.find('Register');
        const button = form.find('button');
        const emailInput = form.find('input#email').at(0);
        
        button.simulate('click');
        expect(form.state('dataSubmitted')).toBeFalsy();

        emailInput.instance().value = "amy@gmail.com";
        emailInput.simulate('change');
        act(() => {
            form.setState({ name: "Amy" });
            form.setState({ password: "" });
        });
        button.simulate('click');
        expect(form.state('dataSubmitted')).toBeFalsy();

        emailInput.instance().value = "amy@gmail.com";
        emailInput.simulate('change');
        act(() => {
            form.setState({ name: "" });
            form.setState({ password: "secret" });
        });
        button.simulate('click');
        expect(form.state('dataSubmitted')).toBeFalsy();
    });
    it('rejects invalid emails', () => {
        const form = component.find('Register');
        act(() => {
            form.setState({ name: "Amy" });
            form.setState({ password: "secret" });
        });
        const emailInput = form.find('input#email').at(0);
        const emails = [
            "",
            "amy@gmail",
            "amy.gmail.com"
        ];
        for (var email of emails) {
            emailInput.instance().value = email;
            emailInput.simulate('change');
            form.find('button').simulate('click');
            expect(form.state('dataSubmitted')).toBeFalsy();
        }
    });
    it('submits on button click', () => {
        const form = component.find('Register');
        const emailInput = form.find('input#email').at(0);
        emailInput.instance().value = "amy@gmail.com";
        emailInput.simulate('change');
        act(() => {
            form.setState({ name: "Amy" });
            form.setState({ password: "secret" });
        });
        form.find('button').simulate('click');
        expect(form.state('dataSubmitted')).toBeTruthy();
    });
    it('submits on Enter in password field', () => {
        const form = component.find('Register');
        const emailInput = form.find('input#email').at(0);
        emailInput.instance().value = "amy@gmail.com";
        emailInput.simulate('change');
        act(() => {
            form.setState({ name: "Amy" });
            form.setState({ password: "secret" });
        });
        form.find('input#password').simulate('keydown', {key: 'Enter'});
        expect(form.state('dataSubmitted')).toBeTruthy();
    });
    it('sends data to server and handles response 200', async (done) => {
        const mockUser = { id: 1, name: "Amy", email: "amy@gmail.com", entries: 0, joined: "" };

        nock('http://localhost:3001')
            //.log(console.log)
            .persist()
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .post('/register', {name: 'Amy', email: 'amy@gmail.com', password: 'secret'})
            .reply(200, mockUser);

        const form = component.find('Register');
        const emailInput = form.find('input#email').at(0);
        emailInput.instance().value = "amy@gmail.com";
        emailInput.simulate('change');
        act(() => {
            form.setState({ name: "Amy" });
            form.setState({ password: "secret" });
        });
        form.find('button').simulate('click');
        await act(async() => {
            await waitUntil(() => form.state('dataReceived') !== null);
        });
        expect(form.state('dataReceived')).toBeTruthy();
        expect(loadUser).toHaveBeenCalledWith(mockUser);
        expect(onRouteChange).toHaveBeenCalledWith('home');
        done();
    });
    it('handles error response from server', async (done) => {
        nock('http://localhost:3001')
            //.log(console.log)
            .persist()
            .defaultReplyHeaders({ 'access-control-allow-origin': '*' })
            .post('/register', {name: 'Amy', email: 'amy@gmail.com', password: 'secret'})
            .reply(400);

        const form = component.find('Register');
        const emailInput = form.find('input#email').at(0);
        emailInput.instance().value = "amy@gmail.com";
        emailInput.simulate('change');
        act(() => {
            form.setState({ name: "Amy" });
            form.setState({ password: "secret" });
        });
        form.find('button').simulate('click');
        await act(async() => {
            await waitUntil(() => form.state('dataReceived') !== null);
        });
        expect(form.state('dataReceived')).toBe(false);
        done();
    });
})
