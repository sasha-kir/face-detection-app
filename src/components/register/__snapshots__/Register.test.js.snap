// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register renders correctly 1`] = `
<ForwardRef(WithAlert(Register))>
  <Register>
    <div
      className="sign-in-main-div center-content"
    >
      <div
        className="sign-in-wrapper"
      >
        <form
          className="register-form sign-in-form"
        >
          <div
            className="form-element f4 white"
          >
            <h1>
              Registration
            </h1>
          </div>
          <div
            className="form-element block-cube block-input"
          >
            <input
              id="name"
              onChange={[Function]}
              placeholder="your name"
              required={true}
              type="text"
            />
            <div
              className="cube-top"
            >
              <div
                className="cube-top-inner"
              />
            </div>
            <div
              className="cube-right"
            >
              <div
                className="cube-right-inner"
              />
            </div>
            <div
              className="cube"
            >
              <div
                className="cube-inner"
              />
            </div>
          </div>
          <div
            className="form-element block-cube block-input"
          >
            <input
              id="email"
              onChange={[Function]}
              pattern="^[a-zA-Z0-9\\\\.\\\\-_]+@[a-z]+\\\\.[a-z]{2,3}$"
              placeholder="email"
              required={true}
              type="text"
            />
            <div
              className="cube-top"
            >
              <div
                className="cube-top-inner"
              />
            </div>
            <div
              className="cube-right"
            >
              <div
                className="cube-right-inner"
              />
            </div>
            <div
              className="cube"
            >
              <div
                className="cube-inner"
              />
            </div>
          </div>
          <div
            className="form-element block-cube block-input"
          >
            <input
              id="password"
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="password"
              required={true}
              type="password"
            />
            <div
              className="cube-top"
            >
              <div
                className="cube-top-inner"
              />
            </div>
            <div
              className="cube-right"
            >
              <div
                className="cube-right-inner"
              />
            </div>
            <div
              className="cube"
            >
              <div
                className="cube-inner"
              />
            </div>
          </div>
          <button
            className="block-button block-cube grow"
            onClick={[Function]}
            type="button"
          >
            <div
              className="cube-top"
            >
              <div
                className="cube-top-inner"
              />
            </div>
            <div
              className="cube-right"
            >
              <div
                className="cube-right-inner"
              />
            </div>
            <div
              className="cube"
            >
              <div
                className="cube-inner"
              />
            </div>
            <div
              className="login-text button-text"
            >
              Sign Up
            </div>
          </button>
          <div
            className="link dim pointer mt4 mb1 f5 white center-content"
            id="signin-link"
            onClick={[Function]}
          >
            Go back
          </div>
        </form>
      </div>
    </div>
  </Register>
</ForwardRef(WithAlert(Register))>
`;
